In [320]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [321]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [322]:
# import excel file StateGeneralbyPrecinctNov2016.xlsx
tn_2016=pd.read_excel('../data/2016_election/StateGeneralbyPrecinctNov2016.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2016

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9397    Wilson       Absentee   
9398    Wilson       Absentee   
9399    Wilson       Absentee   
9400    Wilson       Absentee   
9401    Wilson       Absentee   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9397                            United States President  2016-11-08   
9398                            United States President  2016-11-08   
9399  United States House of Representatives District 6  2016-11-08   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   

                              RNAME1       PARTY1  PVTALLY1  \
0                    Donald J. Trump   Republican      1222   
1     Write-In - Kyle Kenley Kopitke          NaN         0   
2                  Chuck Fleischmann   Republican      1124   
3                   Dennis H. Powers   Republican      1128   
4                    Donald J. Trump   Republican       243   
...                              ...          ...       ...   
9397                 Donald J. Trump   Republican       661   
9398  Write-In - Kyle Kenley Kopitke  Independent         0   
9399                     Diane Black   Republican       699   
9400                       Mark Pody   Republican       264   
9401                    Susan M Lynn   Republican       432   

                             RNAME2       PARTY2  PVTALLY2  \
0                   Hillary Clinton   Democratic     227.0   
1     Write-In - Laurence Kotlikoff          NaN       0.0   
2                    Melody Shekari   Democratic     195.0   
3                        Bob Fannon   Democratic     265.0   
4                   Hillary Clinton   Democratic      42.0   
...                             ...          ...       ...   
9397                Hillary Clinton   Democratic     397.0   
9398  Write-In - Laurence Kotlikoff  Independent       0.0   
9399                  David W. Kent   Democratic     297.0   
9400          Amelia Morrison Hipps   Democratic     142.0   
9401                  Trisha Farmer   Democratic     219.0   

                          RNAME3       PARTY3  PVTALLY3  \
0     "Rocky" Roque De La Fuente  Independent       2.0   
1      Write-In - David Limbaugh          NaN       0.0   
2                Topher Kersting  Independent      21.0   
3                            NaN          NaN       NaN   
4     "Rocky" Roque De La Fuente  Independent       3.0   
...                          ...          ...       ...   
9397  "Rocky" Roque De La Fuente  Independent       4.0   
9398   Write-In - David Limbaugh  Independent       0.0   
9399                  David Ross  Independent      71.0   
9400                         NaN          NaN       NaN   
9401                         NaN          NaN       NaN   

                        RNAME4       PARTY4  PVTALLY4  \
0                 Gary Johnson  Independent      32.0   
1     Write-In - Evan McMullin          NaN       3.0   
2           Cassandra Mitchell  Independent      36.0   
3                          NaN          NaN       NaN   
4                 Gary Johnson  Independent       5.0   
...                        ...          ...       ...   
9397              Gary Johnson  Independent      47.0   
9398  Write-In - Evan McMullin  Independent      14.0   
9399                       NaN    

In [323]:
tn_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9402 entries, 0 to 9401
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      9402 non-null   object 
 1   PRECINCT    9402 non-null   object 
 2   OFFICENAME  9402 non-null   object 
 3   ELECTDATE   9402 non-null   object 
 4   RNAME1      9402 non-null   object 
 5   PARTY1      7561 non-null   object 
 6   PVTALLY1    9402 non-null   int64  
 7   RNAME2      7984 non-null   object 
 8   PARTY2      6088 non-null   object 
 9   PVTALLY2    7984 non-null   float64
 10  RNAME3      5738 non-null   object 
 11  PARTY3      3787 non-null   object 
 12  PVTALLY3    5738 non-null   float64
 13  RNAME4      4913 non-null   object 
 14  PARTY4      2726 non-null   object 
 15  PVTALLY4    4913 non-null   float64
 16  RNAME5      4567 non-null   object 
 17  PARTY5      2726 non-null   object 
 18  PVTALLY5    4567 non-null   float64
 19  RNAME6      2269 non-null  

In [324]:
# merge entries in all RNAME columns into new column "name"
tn_2016_namemelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2016_namemelt=tn_2016_namemelt.drop(['variable'],axis=1)                          
tn_2016_namemelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

                                name  
0                    Donald J. Trump  
1     Write-In - Kyle Kenley Kopitke  
2                  Chuck Fleischmann  
3                   Dennis H. Powers  
4                    Donald J. Trump  
...                              ...  
9400                       Mark Pody  
9401                    Susan M Lynn  
9402                 Hillary Clinton  
9403   Write-In - Laurence Kotlikoff  
9404                  Melody Shekari  

[9405 rows x 5 columns]

In [325]:
# merge entries in all PARTY columns into new column "party"
tn_2016_partymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2016_partymelt=tn_2016_partymelt.drop(['variable'],axis=1)
tn_2016_partymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

           party  
0     Republican  
1            NaN  
2     Republican  
3     Republican  
4     Republican  
...          ...  
9400  Republican  
9401  Republican  
9402  Democratic  
9403         NaN  
9404  Democratic  

[9405 rows x 5 columns]

In [326]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2016_pvtallymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2016_pvtallymelt=tn_2016_pvtallymelt.drop(['variable'],axis=1)
tn_2016_pvtallymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

      vote_count  
0         1222.0  
1            0.0  
2         1124.0  
3         1128.0  
4          243.0  
...          ...  
9400       264.0  
9401       432.0  
9402       227.0  
9403         0.0  
9404       195.0  

[9405 rows x 5 columns]

In [327]:
# add "party" column to tn_2016_namemelt
party=tn_2016_partymelt['party']
tn_2016_namemelt=tn_2016_namemelt.join(party)
tn_2016_namemelt.head(3)

COUNTY       PRECINCT                                         OFFICENAME  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                            United States President   
2  Anderson  Andersonville  United States House of Representatives District 3   

    ELECTDATE                            name       party  
0  2016-11-08                 Donald J. Trump  Republican  
1  2016-11-08  Write-In - Kyle Kenley Kopitke         NaN  
2  2016-11-08               Chuck Fleischmann  Republican

In [328]:
# add "vote_count" column to tn_2016_namemelt
pvtally=tn_2016_pvtallymelt['vote_count']
tn_2016_namemelt=tn_2016_namemelt.join(pvtally)
tn_2016_namemelt.head(3)

COUNTY       PRECINCT                                         OFFICENAME  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                            United States President   
2  Anderson  Andersonville  United States House of Representatives District 3   

    ELECTDATE                            name       party  vote_count  
0  2016-11-08                 Donald J. Trump  Republican      1222.0  
1  2016-11-08  Write-In - Kyle Kenley Kopitke         NaN         0.0  
2  2016-11-08               Chuck Fleischmann  Republican      1124.0

In [329]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      94020 non-null  object 
 1   PRECINCT    94020 non-null  object 
 2   OFFICENAME  94020 non-null  object 
 3   ELECTDATE   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       27969 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [330]:
# lower case column names
tn_2016_namemelt.columns=tn_2016_namemelt.columns.str.lower()
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke         NaN         0.0

In [331]:
# change null values in "party" column to "no party"
tn_2016_namemelt['party'].fillna("No party", inplace = True)
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke    No party         0.0

In [332]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [333]:
# change null values in "vote_count" column to "0"
tn_2016_namemelt['vote_count'].fillna(0)

0        1222.0
1           0.0
2        1124.0
3        1128.0
4         243.0
          ...  
94015       0.0
94016       0.0
94017       0.0
94018       0.0
94019       0.0
Name: vote_count, Length: 94020, dtype: float64

In [334]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


### 43208 non-null names and 43208 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [335]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()]

county             precinct  \
9573   Benton           1 Holladay   
9578   Benton     2 Camden Jr High   
9583   Benton   3 Camden City Hall   
9588   Benton  4 Camden Elementary   
9593   Benton            5 Library   
...       ...                  ...   
94014  Wilson          Provisional   
94016  Wilson             Absentee   
94017  Wilson             Absentee   
94018  Wilson             Absentee   
94019  Wilson             Absentee   

                                              officename   electdate name  \
9573                        Tennessee Senate District 24  2016-11-08  NaN   
9578                        Tennessee Senate District 24  2016-11-08  NaN   
9583                        Tennessee Senate District 24  2016-11-08  NaN   
9588                        Tennessee Senate District 24  2016-11-08  NaN   
9593                        Tennessee Senate District 24  2016-11-08  NaN   
...                                                  ...         ...  ...   
94014     Tennessee House of Representatives District 57  2016-11-08  NaN   
94016                            United States President  2016-11-08  NaN   
94017  United States House of Representatives District 6  2016-11-08  NaN   
94018     Tennessee House of Representatives District 46  2016-11-08  NaN   
94019     Tennessee House of Representatives District 57  2016-11-08  NaN   

          party  vote_count  
9573   No party         NaN  
9578   No party         NaN  
9583   No party         NaN  
9588   No party         NaN  
9593   No party         NaN  
...         ...         ...  
94014  No party         NaN  
94016  No party         NaN  
94017  No party         NaN  
94018  No party         NaN  
94019  No party         NaN  

[50812 rows x 7 columns]

In [336]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50812 entries, 9573 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      50812 non-null  object 
 1   precinct    50812 non-null  object 
 2   officename  50812 non-null  object 
 3   electdate   50812 non-null  object 
 4   name        0 non-null      object 
 5   party       50812 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 3.1+ MB


### Looks safe to drop rows with no votes tallied

In [337]:
# drop rows with no name and vote count
tn_2016_final=tn_2016_namemelt.dropna()

In [338]:
tn_2016_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43208 entries, 0 to 94015
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      43208 non-null  object 
 1   precinct    43208 non-null  object 
 2   officename  43208 non-null  object 
 3   electdate   43208 non-null  object 
 4   name        43208 non-null  object 
 5   party       43208 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 2.6+ MB


In [339]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2016_final=tn_2016_final.astype({'vote_count':'int32'})
tn_2016_final

county       precinct  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson     Briceville   
...         ...            ...   
93997    Wilson           24-1   
94001    Wilson           25-1   
94005    Wilson           25-2   
94010    Wilson    Provisional   
94015    Wilson       Absentee   

                                              officename   electdate  \
0                                United States President  2016-11-08   
1                                United States President  2016-11-08   
2      United States House of Representatives District 3  2016-11-08   
3         Tennessee House of Representatives District 36  2016-11-08   
4                                United States President  2016-11-08   
...                                                  ...         ...   
93997                            United States President  2016-11-08   
94001                            United States President  2016-11-08   
94005                            United States President  2016-11-08   
94010                            United States President  2016-11-08   
94015                            United States President  2016-11-08   

                                 name        party  vote_count  
0                     Donald J. Trump   Republican        1222  
1      Write-In - Kyle Kenley Kopitke     No party           0  
2                   Chuck Fleischmann   Republican        1124  
3                    Dennis H. Powers   Republican        1128  
4                     Donald J. Trump   Republican         243  
...                               ...          ...         ...  
93997         Write-In - Tom Hoefling  Independent           0  
94001         Write-In - Tom Hoefling  Independent           0  
94005         Write-In - Tom Hoefling  Independent           0  
94010         Write-In - Tom Hoefling  Independent           0  
94015         Write-In - Tom Hoefling  Independent           0  

[43208 rows x 7 columns]

In [340]:
tn_2016_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [341]:
tn_2016_final['officename'].value_counts()

United States President                              30330
United States House of Representatives District 8     1729
United States House of Representatives District 3     1380
United States House of Representatives District 1      868
United States House of Representatives District 7      852
                                                     ...  
Tennessee House of Representatives District 99          12
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 68          11
Tennessee House of Representatives District 24          10
Tennessee House of Representatives District 16          10
Name: officename, Length: 125, dtype: int64

In [342]:
tn_2016_final['officename'].unique()

array(['United States President',
       'United States House of Representatives District 3',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee Senate District 14',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee Senate District 24',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee Senate District 10',
       'Tennessee House of Representatives District 24',
       'Tennessee House of Representatives District 22',

In [343]:
# create data frame tn_2016_pres for presidental election
tn_2016_pres=tn_2016_final[tn_2016_final['officename'].str.match('United States President')]
tn_2016_pres=tn_2016_pres.reset_index()
tn_2016_pres

index    county       precinct               officename   electdate  \
0          0  Anderson  Andersonville  United States President  2016-11-08   
1          1  Anderson  Andersonville  United States President  2016-11-08   
2          4  Anderson     Briceville  United States President  2016-11-08   
3          5  Anderson     Briceville  United States President  2016-11-08   
4          8  Anderson       Bull Run  United States President  2016-11-08   
...      ...       ...            ...                      ...         ...   
30325  93997    Wilson           24-1  United States President  2016-11-08   
30326  94001    Wilson           25-1  United States President  2016-11-08   
30327  94005    Wilson           25-2  United States President  2016-11-08   
30328  94010    Wilson    Provisional  United States President  2016-11-08   
30329  94015    Wilson       Absentee  United States President  2016-11-08   

                                 name        party  vote_count  
0                     Donald J. Trump   Republican        1222  
1      Write-In - Kyle Kenley Kopitke     No party           0  
2                     Donald J. Trump   Republican         243  
3      Write-In - Kyle Kenley Kopitke     No party           0  
4                     Donald J. Trump   Republican        1397  
...                               ...          ...         ...  
30325         Write-In - Tom Hoefling  Independent           0  
30326         Write-In - Tom Hoefling  Independent           0  
30327         Write-In - Tom Hoefling  Independent           0  
30328         Write-In - Tom Hoefling  Independent           0  
30329         Write-In - Tom Hoefling  Independent           0  

[30330 rows x 8 columns]

In [344]:
tn_2016_pres['vote_count'].value_counts()

0       16240
1        2257
2        1548
3        1074
4         805
        ...  
1959        1
1927        1
1951        1
1831        1
1847        1
Name: vote_count, Length: 1484, dtype: int64

### 16240 rows have a vote count of 0. Drop those rows

In [345]:
tn_2016_pres=tn_2016_pres[tn_2016_pres['vote_count'] != 0]
tn_2016_pres=tn_2016_pres.reset_index()

In [346]:
# group county precincts into county per candidate 
tn_2016_pres=tn_2016_pres.groupby(['officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2016_pres

officename    county                           name  \
0    United States President  Anderson     "Rocky" Roque De La Fuente   
1    United States President  Anderson                 Alyson Kennedy   
2    United States President  Anderson                Donald J. Trump   
3    United States President  Anderson                   Gary Johnson   
4    United States President  Anderson                Hillary Clinton   
..                       ...       ...                            ...   
884  United States President    Wilson                     Mike Smith   
885  United States President    Wilson   Write-In - Darrell L. Castle   
886  United States President    Wilson       Write-In - Evan McMullin   
887  United States President    Wilson  Write-In - Laurence Kotlikoff   
888  United States President    Wilson        Write-In - Tom Hoefling   

           party  vote_count  
0    Independent          74  
1    Independent          42  
2     Republican       19212  
3    Independent        1034  
4     Democratic        9013  
..           ...         ...  
884  Independent         207  
885  Independent          23  
886  Independent         280  
887  Independent           1  
888  Independent           6  

[889 rows x 5 columns]

In [347]:
tn_2016_pres.head(20)

officename    county                          name  \
0   United States President  Anderson    "Rocky" Roque De La Fuente   
1   United States President  Anderson                Alyson Kennedy   
2   United States President  Anderson               Donald J. Trump   
3   United States President  Anderson                  Gary Johnson   
4   United States President  Anderson               Hillary Clinton   
5   United States President  Anderson                    Jill Stein   
6   United States President  Anderson                    Mike Smith   
7   United States President  Anderson  Write-In - Darrell L. Castle   
8   United States President  Anderson      Write-In - Evan McMullin   
9   United States President   Bedford    "Rocky" Roque De La Fuente   
10  United States President   Bedford                Alyson Kennedy   
11  United States President   Bedford               Donald J. Trump   
12  United States President   Bedford                  Gary Johnson   
13  United States President   Bedford               Hillary Clinton   
14  United States President   Bedford                    Jill Stein   
15  United States President   Bedford                    Mike Smith   
16  United States President   Bedford  Write-In - Darrell L. Castle   
17  United States President   Bedford      Write-In - Evan McMullin   
18  United States President    Benton    "Rocky" Roque De La Fuente   
19  United States President    Benton                Alyson Kennedy   

          party  vote_count  
0   Independent          74  
1   Independent          42  
2    Republican       19212  
3   Independent        1034  
4    Democratic        9013  
5   Independent         189  
6   Independent         174  
7      No party          28  
8      No party         115  
9   Independent          25  
10  Independent           8  
11   Republican       11486  
12  Independent         306  
13   Democratic        3395  
14  Independent          75  
15  Independent          47  
16     No party           5  
17     No party           8  
18  Independent          12  
19  Independent          13

### Add column that has total vote count by count for use in calculating county vote percentages

In [348]:
tn_2016_pres['sum_county_vote']=tn_2016_pres.groupby(['vote_count'])['vote_count'].sum
tn_2016_pres

officename    county                           name  \
0    United States President  Anderson     "Rocky" Roque De La Fuente   
1    United States President  Anderson                 Alyson Kennedy   
2    United States President  Anderson                Donald J. Trump   
3    United States President  Anderson                   Gary Johnson   
4    United States President  Anderson                Hillary Clinton   
..                       ...       ...                            ...   
884  United States President    Wilson                     Mike Smith   
885  United States President    Wilson   Write-In - Darrell L. Castle   
886  United States President    Wilson       Write-In - Evan McMullin   
887  United States President    Wilson  Write-In - Laurence Kotlikoff   
888  United States President    Wilson        Write-In - Tom Hoefling   

           party  vote_count  \
0    Independent          74   
1    Independent          42   
2     Republican       19212   
3    Independent        1034   
4     Democratic        9013   
..           ...         ...   
884  Independent         207   
885  Independent          23   
886  Independent         280   
887  Independent           1   
888  Independent           6   

                                       sum_county_vote  
0    <bound method GroupBy.sum of <pandas.core.grou...  
1    <bound method GroupBy.sum of <pandas.core.grou...  
2    <bound method GroupBy.sum of <pandas.core.grou...  
3    <bound method GroupBy.sum of <pandas.core.grou...  
4    <bound method GroupBy.sum of <pandas.core.grou...  
..                                                 ...  
884  <bound method GroupBy.sum of <pandas.core.grou...  
885  <bound method GroupBy.sum of <pandas.core.grou...  
886  <bound method GroupBy.sum of <pandas.core.grou...  
887  <bound method GroupBy.sum of <pandas.core.grou...  
888  <bound method GroupBy.sum of <pandas.core.grou...  

[889 rows x 6 columns]

In [349]:
filter=tn_2016_pres['county']==tn_2016_pres['county']
tn_2016_pres['sum_county_vote']=tn_2016_pres.where(filter).sum
tn_2016_pres

RecursionError: maximum recursion depth exceeded while calling a Python object

RecursionError: maximum recursion depth exceeded while calling a Python object

In [350]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2016_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0    Anderson          74
1    Anderson          42
2    Anderson       19212
3    Anderson        1034
4    Anderson        9013
..        ...         ...
884    Wilson         207
885    Wilson          23
886    Wilson         280
887    Wilson           1
888    Wilson           6

[889 rows x 2 columns]

In [351]:
# sum votes per county
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       29881
Bedford        15355
Benton          6328
Bledsoe         4664
Blount         52208
               ...  
Wayne           5857
Weakley        12184
White           9825
Williamson    106271
Wilson         56734
Name: vote_count, Length: 95, dtype: int32

In [352]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2016_pres['sum_county_vote']=tn_2016_pres['county'].map(county_vote_sum)
tn_2016_pres.head(20)

officename    county                          name  \
0   United States President  Anderson    "Rocky" Roque De La Fuente   
1   United States President  Anderson                Alyson Kennedy   
2   United States President  Anderson               Donald J. Trump   
3   United States President  Anderson                  Gary Johnson   
4   United States President  Anderson               Hillary Clinton   
5   United States President  Anderson                    Jill Stein   
6   United States President  Anderson                    Mike Smith   
7   United States President  Anderson  Write-In - Darrell L. Castle   
8   United States President  Anderson      Write-In - Evan McMullin   
9   United States President   Bedford    "Rocky" Roque De La Fuente   
10  United States President   Bedford                Alyson Kennedy   
11  United States President   Bedford               Donald J. Trump   
12  United States President   Bedford                  Gary Johnson   
13  United States President   Bedford               Hillary Clinton   
14  United States President   Bedford                    Jill Stein   
15  United States President   Bedford                    Mike Smith   
16  United States President   Bedford  Write-In - Darrell L. Castle   
17  United States President   Bedford      Write-In - Evan McMullin   
18  United States President    Benton    "Rocky" Roque De La Fuente   
19  United States President    Benton                Alyson Kennedy   

          party  vote_count  sum_county_vote  
0   Independent          74            29881  
1   Independent          42            29881  
2    Republican       19212            29881  
3   Independent        1034            29881  
4    Democratic        9013            29881  
5   Independent         189            29881  
6   Independent         174            29881  
7      No party          28            29881  
8      No party         115            29881  
9   Independent          25            15355  
10  Independent           8            15355  
11   Republican       11486            15355  
12  Independent         306            15355  
13   Democratic        3395            15355  
14  Independent          75            15355  
15  Independent          47            15355  
16     No party           5            15355  
17     No party           8            15355  
18  Independent          12             6328  
19  Independent          13             6328

In [353]:
# create "pct_county_vote" column for tn_2016_pres
tn_2016_pres['pct_county_vote']=round(tn_2016_pres['vote_count']/tn_2016_pres['sum_county_vote']*100,2)
tn_2016_pres

officename    county                           name  \
0    United States President  Anderson     "Rocky" Roque De La Fuente   
1    United States President  Anderson                 Alyson Kennedy   
2    United States President  Anderson                Donald J. Trump   
3    United States President  Anderson                   Gary Johnson   
4    United States President  Anderson                Hillary Clinton   
..                       ...       ...                            ...   
884  United States President    Wilson                     Mike Smith   
885  United States President    Wilson   Write-In - Darrell L. Castle   
886  United States President    Wilson       Write-In - Evan McMullin   
887  United States President    Wilson  Write-In - Laurence Kotlikoff   
888  United States President    Wilson        Write-In - Tom Hoefling   

           party  vote_count  sum_county_vote  pct_county_vote  
0    Independent          74            29881             0.25  
1    Independent          42            29881             0.14  
2     Republican       19212            29881            64.30  
3    Independent        1034            29881             3.46  
4     Democratic        9013            29881            30.16  
..           ...         ...              ...              ...  
884  Independent         207            56734             0.36  
885  Independent          23            56734             0.04  
886  Independent         280            56734             0.49  
887  Independent           1            56734             0.00  
888  Independent           6            56734             0.01  

[889 rows x 7 columns]

In [354]:
# how many parties?
tn_2016_pres['party'].value_counts()

Independent    504
No party       195
Democratic      95
Republican      95
Name: party, dtype: int64

### Create "rep_vic_margin" column  with percent value difference for Republican compared to other parties for heat mapping

In [355]:
# create dataframe "tn_2016_margin"
tn_2016_pct_party_county=tn_2016_pres.loc[0:,['county','party','pct_county_vote']]
tn_2016_margin

county      party_x  pct_county_vote     party_y  rep_pct_x  \
0    Anderson  Independent             0.25  Republican      64.30   
1    Anderson  Independent             0.14  Republican      64.30   
2    Anderson   Republican            64.30  Republican      64.30   
3    Anderson  Independent             3.46  Republican      64.30   
4    Anderson   Democratic            30.16  Republican      64.30   
..        ...          ...              ...         ...        ...   
884    Wilson  Independent             0.36  Republican      69.46   
885    Wilson  Independent             0.04  Republican      69.46   
886    Wilson  Independent             0.49  Republican      69.46   
887    Wilson  Independent             0.00  Republican      69.46   
888    Wilson  Independent             0.01  Republican      69.46   

     rep_vic_margin       party  rep_pct_y  
0             64.05  Republican      64.30  
1             64.16  Republican      64.30  
2              0.00  Republican      64.30  
3             60.84  Republican      64.30  
4             34.14  Republican      64.30  
..              ...         ...        ...  
884           69.10  Republican      69.46  
885           69.42  Republican      69.46  
886           68.97  Republican      69.46  
887           69.46  Republican      69.46  
888           69.45  Republican      69.46  

[889 rows x 8 columns]

In [356]:
# create new dataframe with republican values only
tn_2016_rep_pct=tn_2016_pct_party_county[tn_2016_pct_party_county['party']== "Republican"].reset_index(drop=True)
tn_2016_rep_pct

county       party  pct_county_vote
0     Anderson  Republican            64.30
1      Bedford  Republican            74.80
2       Benton  Republican            74.53
3      Bledsoe  Republican            77.66
4       Blount  Republican            71.72
..         ...         ...              ...
90       Wayne  Republican            85.98
91     Weakley  Republican            73.93
92       White  Republican            78.08
93  Williamson  Republican            64.19
94      Wilson  Republican            69.46

[95 rows x 3 columns]

In [357]:
# rename 'pct_county_vote' to 'rep_pct' to prepare for merge 
tn_2016_rep_pct=tn_2016_rep_pct.rename(columns={'pct_county_vote':'rep_pct'},inplace=False)
tn_2016_rep_pct.head(2)

county       party  rep_pct
0  Anderson  Republican     64.3
1   Bedford  Republican     74.8

In [358]:
# use left merge to tn_2016_margin of tn_2016_rep_pct to fill in appropriate county values for Republican percentage
tn_2016_margin=tn_2016_margin.merge(tn_2016_rep_pct,on='county')
tn_2016_margin.head(20)

county      party_x  pct_county_vote     party_y  rep_pct_x  \
0   Anderson  Independent             0.25  Republican      64.30   
1   Anderson  Independent             0.14  Republican      64.30   
2   Anderson   Republican            64.30  Republican      64.30   
3   Anderson  Independent             3.46  Republican      64.30   
4   Anderson   Democratic            30.16  Republican      64.30   
5   Anderson  Independent             0.63  Republican      64.30   
6   Anderson  Independent             0.58  Republican      64.30   
7   Anderson     No party             0.09  Republican      64.30   
8   Anderson     No party             0.38  Republican      64.30   
9    Bedford  Independent             0.16  Republican      74.80   
10   Bedford  Independent             0.05  Republican      74.80   
11   Bedford   Republican            74.80  Republican      74.80   
12   Bedford  Independent             1.99  Republican      74.80   
13   Bedford   Democratic            22.11  Republican      74.80   
14   Bedford  Independent             0.49  Republican      74.80   
15   Bedford  Independent             0.31  Republican      74.80   
16   Bedford     No party             0.03  Republican      74.80   
17   Bedford     No party             0.05  Republican      74.80   
18    Benton  Independent             0.19  Republican      74.53   
19    Benton  Independent             0.21  Republican      74.53   

    rep_vic_margin     party_x  rep_pct_y     party_y  rep_pct  
0            64.05  Republican      64.30  Republican    64.30  
1            64.16  Republican      64.30  Republican    64.30  
2             0.00  Republican      64.30  Republican    64.30  
3            60.84  Republican      64.30  Republican    64.30  
4            34.14  Republican      64.30  Republican    64.30  
5            63.67  Republican      64.30  Republican    64.30  
6            63.72  Republican      64.30  Republican    64.30  
7            64.21  Republican      64.30  Republican    64.30  
8            63.92  Republican      64.30  Republican    64.30  
9            74.64  Republican      74.80  Republican    74.80  
10           74.75  Republican      74.80  Republican    74.80  
11            0.00  Republican      74.80  Republican    74.80  
12           72.81  Republican      74.80  Republican    74.80  
13           52.69  Republican      74.80  Republican    74.80  
14           74.31  Republican      74.80  Republican    74.80  
15           74.49  Republican      74.80  Republican    74.80  
16           74.77  Republican      74.80  Republican    74.80  
17           74.75  Republican      74.80  Republican    74.80  
18           74.34  Republican      74.53  Republican    74.53  
19           74.32  Republican      74.53  Republican    74.53

In [359]:
tn_2016_margin

county      party_x  pct_county_vote     party_y  rep_pct_x  \
0    Anderson  Independent             0.25  Republican      64.30   
1    Anderson  Independent             0.14  Republican      64.30   
2    Anderson   Republican            64.30  Republican      64.30   
3    Anderson  Independent             3.46  Republican      64.30   
4    Anderson   Democratic            30.16  Republican      64.30   
..        ...          ...              ...         ...        ...   
884    Wilson  Independent             0.36  Republican      69.46   
885    Wilson  Independent             0.04  Republican      69.46   
886    Wilson  Independent             0.49  Republican      69.46   
887    Wilson  Independent             0.00  Republican      69.46   
888    Wilson  Independent             0.01  Republican      69.46   

     rep_vic_margin     party_x  rep_pct_y     party_y  rep_pct  
0             64.05  Republican      64.30  Republican    64.30  
1             64.16  Republican      64.30  Republican    64.30  
2              0.00  Republican      64.30  Republican    64.30  
3             60.84  Republican      64.30  Republican    64.30  
4             34.14  Republican      64.30  Republican    64.30  
..              ...         ...        ...         ...      ...  
884           69.10  Republican      69.46  Republican    69.46  
885           69.42  Republican      69.46  Republican    69.46  
886           68.97  Republican      69.46  Republican    69.46  
887           69.46  Republican      69.46  Republican    69.46  
888           69.45  Republican      69.46  Republican    69.46  

[889 rows x 10 columns]

In [360]:
# add 'rep_vic_margin' column to show Republican margin over other parties per county
tn_2016_margin['repub_vic_margin']=tn_2016_margin['rep_pct']-tn_2016_margin['pct_county_vote']
tn_2016_margin

county      party_x  pct_county_vote     party_y  rep_pct_x  \
0    Anderson  Independent             0.25  Republican      64.30   
1    Anderson  Independent             0.14  Republican      64.30   
2    Anderson   Republican            64.30  Republican      64.30   
3    Anderson  Independent             3.46  Republican      64.30   
4    Anderson   Democratic            30.16  Republican      64.30   
..        ...          ...              ...         ...        ...   
884    Wilson  Independent             0.36  Republican      69.46   
885    Wilson  Independent             0.04  Republican      69.46   
886    Wilson  Independent             0.49  Republican      69.46   
887    Wilson  Independent             0.00  Republican      69.46   
888    Wilson  Independent             0.01  Republican      69.46   

     rep_vic_margin     party_x  rep_pct_y     party_y  rep_pct  \
0             64.05  Republican      64.30  Republican    64.30   
1             64.16  Republican      64.30  Republican    64.30   
2              0.00  Republican      64.30  Republican    64.30   
3             60.84  Republican      64.30  Republican    64.30   
4             34.14  Republican      64.30  Republican    64.30   
..              ...         ...        ...         ...      ...   
884           69.10  Republican      69.46  Republican    69.46   
885           69.42  Republican      69.46  Republican    69.46   
886           68.97  Republican      69.46  Republican    69.46   
887           69.46  Republican      69.46  Republican    69.46   
888           69.45  Republican      69.46  Republican    69.46   

     repub_vic_margin  
0               64.05  
1               64.16  
2                0.00  
3               60.84  
4               34.14  
..                ...  
884             69.10  
885             69.42  
886             68.97  
887             69.46  
888             69.45  

[889 rows x 11 columns]

In [361]:
# Confirm negative numbers
tn_2016_margin['rep_vic_margin'].unique()

array([ 64.05,  64.16,   0.  ,  60.84,  34.14,  63.67,  63.72,  64.21,
        63.92,  74.64,  74.75,  72.81,  52.69,  74.31,  74.49,  74.77,
        74.34,  74.32,  73.22,  51.24,  74.36,  74.5 ,  74.48,  77.51,
        77.57,  75.62,  58.43,  77.27,  77.47,  77.62,  77.45,  71.56,
        71.58,  68.53,  48.54,  71.12,  71.24,  71.72,  71.62,  71.29,
        76.82,  76.88,  74.15,  58.71,  76.45,  76.62,  76.99,  76.91,
        76.38,  78.5 ,  78.56,  76.76,  60.82,  78.4 ,  78.42,  78.72,
        78.66,  75.25,  75.27,  73.41,  54.19,  74.85,  75.06,  75.34,
        75.23,  72.95,  52.28,  74.52,  74.67,  74.51,  80.01,  80.06,
        78.21,  63.77,  79.53,  79.89,  80.03,  79.85,  70.78,  70.86,
        68.03,  46.59,  70.32,  70.63,  70.89,  70.37,  78.01,  78.03,
        76.32,  58.99,  77.94,  77.91,  77.58,  79.94,  79.87,  78.3 ,
        63.03,  79.82,  79.84,  80.07,  73.25,  71.71,  49.11,  72.94,
        72.98,  80.57,  80.58,  79.  ,  64.4 ,  80.25,  80.52,  80.72,
      

In [362]:
# join tn_2016_margin 'rep_vic_margin' column to tn_2016_pres
rep_pct=tn_2016_margin['repub_vic_margin']
tn_2016_pres=tn_2016_pres.join(rep_pct)
tn_2016_pres.head(3)

officename    county                        name        party  \
0  United States President  Anderson  "Rocky" Roque De La Fuente  Independent   
1  United States President  Anderson              Alyson Kennedy  Independent   
2  United States President  Anderson             Donald J. Trump   Republican   

   vote_count  sum_county_vote  pct_county_vote  repub_vic_margin  
0          74            29881             0.25             64.05  
1          42            29881             0.14             64.16  
2       19212            29881            64.30              0.00

In [364]:
#zip_cbsa.to_excel(r'../data/zip_cbsa.xlsx',index=False)
tn_2016_pres.to_csv(r'../data/2016_election/tn_2016_pres.csv',index=False)

In [365]:
tn_2016_pres.to_excel(r'../data/2016_election/tn_2016_pres.xlsx',index=False)